### PR TITLE
Groups

### PR DESCRIPTION
**Pull request by IanPhilips**
This is now in a state to test! We've got:
- groups that contain contracts that only members of the group can add to the groups page. The creator or added members can add other members to the group to give them the appropriate permission.
- creating new groups from  /groups and /create
- adding members during group creation and afterwards via editing the group
- autonaming a group via its members so that you don't need to add a name
- slight nod to groups on a contract page if the question belongs to a group - this needs work and I'd love input on how to highlight a question belongs to a group!
- no group discussion wall yet! Will the new and mysterious `Message.ts` work for this?
- notifications when you're added to a group
- groups tab on your profile
- groups icons on navigation - this should be more thought out and improved upon!

### CODE REVIEW

**Comment by IanPhilips** [comment]
Hm, I really do want to be able to follow a group if I can't join it.

**Comment by IanPhilips** [comment]
I suppose I can follow the creators for now. If large groups form though I'll want to follow a group I think.

**Comment by IanPhilips** [comment]
Part of the reason people may disagree with the design choices is the name group is so ambiguous. I think Channel captures the single organized site of information production part of the group functionality.

**Comment by jahooma** [comment]
This is really cool! I think groups will be a lot of fun.

If people can reach the groups through notifications (comment in group discussion, new market in group) then that will solve a lot of the problems of people just never navigating there. (And it gives value to joining random groups that are mostly inactive.) Plus we can show them in the sidebar like you have.

- I think the overall structure of the group makes sense, provided we add the discussion tab!
- I like how you can add people to groups. (And the selector is very slick!)
- The My groups vs All groups tabs are nice!
- What do you think about adding existing markets to the group? Maybe some kind of search and select widget?
- Channel vs. Group may become clear after we see how people use them. What use cases do people really want?

The only other criticism I have are that the UI is still a bit rough in a few places.
 
- Long group names don't fit very well.
Screen Shot 2022-06-15 at 9 20 07 PM
- Names a bit cut off
Screen Shot 2022-06-15 at 9 48 08 PM
- Group card formatted weird on mobile
Screen Shot 2022-06-15 at 9 51 40 PM

I'll take a look at the code in a bit.

**Comment by IanPhilips** [comment]
Thanks for looking at the UI, James! I wouldn't look at the code until
tomorrow after I'm done with messages, though. I haven't done any
proofreading yet!

**Review by jahooma — COMMENTED** [review]
Nice work here! Really cool that you could jump off from all the folds code.

Also, wow, this is a gigantic PR! I scrolled through a lot of the UI code — can take a more in depth look in another pass.

One generic suggestion is that if you could extract some of the Tailwind UI code into reusable components, that would be great! It helps us in the future, but it also makes the code a lot more readable IMO.

Also, it's hard to see groups as both the private group chat with friends and the market collection use case of tags and categories. I'm not certain about removing categories, if groups become more like an expansion of the Personal category instead of a replacement for broad science / technology / economics etc. Public vs private markets. Global vs local. Hmm...

File: firestore.rules [hunk 1]
```diff
@@ -111,5 +111,13 @@ service cloud.firestore {
                        && request.resource.data.diff(resource.data).affectedKeys()
                                                                     .hasOnly(['isSeen', 'viewTime']);
     }
+
+    match /groups/{groupId} {
+          allow read;
+          allow update: if request.auth.uid == resource.data.curatorId
```
[the remark this answers is not in the document]
Comment (IanPhilips) [reply firestore.rules:117]: Nice, thanks for catching!

[the remark this answers is not in the document]
Comment (IanPhilips) [reply firestore.rules:117]: Oh I had updated them in the console! just hadn't copied them. Good eye anyways though

File: common/message.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
+export type Message = {
```
[the remark this answers is not in the document]
Comment (IanPhilips) [reply common/message.ts:1]: Hmmm I suppose we can! I was hesitant to change the optionality of contractId on Comment but that would be better overall so I'll give that a go. Thanks for the rec!

File: functions/src/on-update-group.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  .onUpdate(async (change) => {
+    const prevGroup = change.before.data() as Group
+    const group = change.after.data() as Group
+
+    // ignore the update we just made
+    if (prevGroup.mostRecentActivityTime !== group.mostRecentActivityTime)
+      return
+
+    group.mostRecentActivityTime = new Date().getTime()
+    await firestore.collection('groups').doc(group.id).set(group)
```
[the remark this answers is not in the document]
Comment (IanPhilips) [reply functions/src/on-update-group.ts:17]: Sounds good

File: functions/src/create-group.ts [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+      return {
+        status: 'error',
+        message: 'memberIds must be an array of strings',
+      }
+
+    if (typeof anyoneCanJoin !== 'boolean')
+      return { status: 'error', message: 'anyoneCanJoin must be a boolean' }
+
+    if (typeof visibility !== 'string')
+      return { status: 'error', message: 'visibility must be a string' }
```
[the remark this answers is not in the document]
Comment (IanPhilips) [reply functions/src/create-group.ts:48]: good idea! That makes validation much cleaner.

File: functions/src/create-group.ts [hunk 5]
```diff
@@ -0,0 +1,113 @@
+import * as functions from 'firebase-functions'
+import * as admin from 'firebase-admin'
+
+import { getUser } from './utils'
+import { Contract } from '../../common/contract'
+import { slugify } from '../../common/util/slugify'
+import { randomString } from '../../common/util/random'
+import { Group } from '../../common/group'
+
+export const createGroup = functions.runWith({ minInstances: 1 }).https.onCall(
```
[the remark this answers is not in the document]
Comment (IanPhilips) [reply functions/src/create-group.ts:10]: Cool! Thanks for the suggestion, I'll try that.

File: functions/src/on-group-follow.ts [hunk 6]
```diff
@@ -3,15 +3,15 @@ import * as admin from 'firebase-admin'
 
 const firestore = admin.firestore()
 
-export const onFoldFollow = functions.firestore
-  .document('folds/{foldId}/followers/{userId}')
+export const onGroupFollow = functions.firestore
```
[the remark this answers is not in the document]
Comment (IanPhilips) [reply functions/src/on-group-follow.ts:6]: I'll delete the follow stuff for now

**Comment by IanPhilips** [comment]
Allowing users to add old questions to a group:

Screen Shot 2022-06-17 at 2 42 39 PM

Right now I limit the number of groups a question can be tied to to one. But it should be fairly trivial to allow multiple groups per question, it will just be a matter of displaying all the groups the question is in.

**Comment by IanPhilips** [comment]
Shoot also I need to hide the comment input box if you're not a member and
figure out why the firestore rules aren't preventing non-members from
making comments.

**Comment by IanPhilips** [comment]
Screen Shot 2022-06-20 at 7 46 05 AM

Non-members won't see a comment input for discussion. Should they? Also the firestore rules aren't preventing non members from adding comments. Actually even if I don't allow anyone to create a coment they still don't work so I'd love input on that.

File: web/pages/group/[...slugs]/index.tsx [hunk 7]
```diff
@@ -0,0 +1,485 @@
...
+
+  async function addContractToGroup(contract: Contract) {
+    await updateGroup(group, {
+      ...group,
+      contractIds: [...group.contractIds, contract.id],
+    })
+    setOpen(false)
+  }
+
+  // List groups with the highest question count, then highest member count
```
Comment (jahooma) [inline web/pages/group/[...slugs]/index.tsx:430]: Remove?

File: web/pages/groups.tsx [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+          <Col className="px-4 sm:px-0">
+            <Row className="items-center justify-between">
+              <Title text="Explore groups" />
+              {user && (
+                <CreateGroupButton user={user} goToGroupOnSubmit={true} />
+              )}
+            </Row>
+
+            <div className="mb-6 text-gray-500">
+              Groups enable users to curate, discuss, and rank in a subset of
```
Comment (jahooma) [inline web/pages/groups.tsx:103]: Discuss and compete on questions with a group of friends.

File: web/pages/groups.tsx [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+                            group={group}
+                            creator={creatorsDict[group.creatorId]}
+                          />
+                        ))}
+                      </Col>
+                    </Col>
+                  ),
+                },
+              ]
+                .concat(
```
Comment (jahooma) [inline web/pages/groups.tsx:133]: You could do
```
tabs={[
...user ? [{ // My Groups }] : [],
{ // All },
]}
```

File: web/pages/groups.tsx [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+          <div className="mr-1">Created by</div>
+          <UserLink
+            className="text-neutral"
+            name={creator?.name ?? ''}
+            username={creator?.username ?? ''}
+          />
+        </Row>
+        {group.memberIds.length > 1 && (
+          <Row>
+            <GroupMembersList group={group} />
```
Comment (jahooma) [inline web/pages/groups.tsx:199]: Abbreviate this for large groups.

File: web/pages/create.tsx [hunk 11]
```diff
@@ -245,24 +264,12 @@ export function NewContract(props: { question: string }) {
...
-          onChange={(e) => setCategory(e.currentTarget.value ?? '')}
-        >
-          <option value={''}>(none)</option>
-          {Object.entries(CATEGORIES).map(([id, name]) => (
-            <option key={id} value={id}>
-              {name}
-            </option>
-          ))}
-        </select>
-      </div>
```
Comment (jahooma) [inline web/pages/create.tsx:265]: Let's keep this for now. Can make the decision to remove later

**Comment by IanPhilips** [comment]
Screen Shot 2022-06-20 at 12 32 49 PM

**Review by jahooma — APPROVED** [review]
Nice work! I'm ready to start using some groups!

File: web/components/feed/feed-comments.tsx [hunk 12]
```diff
@@ -423,71 +424,120 @@ export function CommentInput(props: {
...
-            </Row>
-            <Row>
-              {!user && (
-                <button
-                  className={'btn btn-outline btn-sm mt-2 normal-case'}
-                  onClick={() => submitComment(id)}
-                >
-                  Sign in to comment
-                </button>
+  const memoizedSetComment = useMemo(() => setComment, [setComment])
```
[the remark this answers is not in the document]
Comment (IanPhilips) [reply web/components/feed/feed-comments.tsx:470]: They don't include the functions in dep array in the examples here: https://reactjs.org/docs/hooks-reference.html#usememo so I figured useMemo would work but I see I DID include the function in the dep array b/c of the linter so I see I was using that wrong. Thanks for the tip!

File: web/components/filter-select-users.tsx [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+import { Row } from 'web/components/layout/row'
+
+export function FilterSelectUsers(props: {
+  setSelectedUsers: (users: User[]) => void
+  ignoreUserIds: string[]
+}) {
+  const { ignoreUserIds } = props
+  const users = useUsers()
+  const [query, setQuery] = useState('')
+  const [selectedUsers, setSelectedUsers] = useState<User[]>([])
```
[the remark this answers is not in the document]
Comment (IanPhilips) [reply web/components/filter-select-users.tsx:17]: Great idea! Much cleaner.

File: web/hooks/use-group.ts [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+        setMemberGroupIds(groupIds)
+        localStorage.setItem(key, JSON.stringify(groupIds))
+      })
+    }
+  }, [user])
+
+  return memberGroupIds
+}
+
+export function useMembers(group: Group) {
```
[the remark this answers is not in the document]
Comment (IanPhilips) [reply web/hooks/use-group.ts:64]: Nice, I should learn it as well and integrate it!

File: web/components/contract/contract-card.tsx [hunk 15]
```diff
@@ -42,12 +44,14 @@ export function ContractCard(props: {
...
-    (outcomeType === 'FREE_RESPONSE' && getTopAnswer(contract) === undefined) ||
-    outcomeType === 'NUMERIC'
-  )
+  const showQuickBet =
+    !(
+      !user ||
+      marketClosed ||
+      (outcomeType === 'FREE_RESPONSE' &&
+        getTopAnswer(contract) === undefined) ||
+      outcomeType === 'NUMERIC'
```
[the remark this answers is not in the document]
Comment (IanPhilips) [reply web/components/contract/contract-card.tsx:53]: Nice, that's cleaner!

File: web/components/contract/contract-info-dialog.tsx [hunk 16]
```diff
@@ -147,9 +146,7 @@ export function ContractInfoDialog(props: { contract: Contract; bets: Bet[] }) {
             </tbody>
           </table>
 
-          <div>Tags</div>
-          <TagsInput contract={contract} />
-          <div />
+          {/*TODO: allow users to add current markets to a group?*/}
```
[the remark this answers is not in the document]
Comment (IanPhilips) [reply web/components/contract/contract-info-dialog.tsx:150]: Will do!

File: web/components/contract/contracts-list.tsx [hunk 17]
```diff
@@ -6,15 +6,26 @@ import { ContractCard } from './contract-card'
...
 import { useEffect, useState } from 'react'
+import clsx from 'clsx'
 
 export function ContractsGrid(props: {
   contracts: Contract[]
   loadMore: () => void
   hasMore: boolean
   showCloseTime?: boolean
   onContractClick?: (contract: Contract) => void
+  className?: string
```
[the remark this answers is not in the document]
Comment (IanPhilips) [reply web/components/contract/contracts-list.tsx:17]: Yes, that's more explicit

File: web/components/groups/discussion.tsx [hunk 18]
```diff
@@ -0,0 +1,186 @@
...
+              enterToSubmit={true}
+            />
+          </div>
+        </div>
+      )}
+    </Col>
+  )
+}
+
+function GroupMessage_(props: {
```
[the remark this answers is not in the document]
Comment (IanPhilips) [reply web/components/groups/discussion.tsx:127]: good idea!

File: web/components/user-page.tsx [hunk 19]
```diff
@@ -287,6 +299,19 @@ export function UserPage(props: {
                   <div className="px-0.5 font-bold">{usersBets.length}</div>
                 ),
               },
+              {
+                title: 'Groups',
```
[the remark this answers is not in the document]
Comment (IanPhilips) [reply web/components/user-page.tsx:303]: Will do! Good eye

File: web/pages/group/[...slugs]/index.tsx (hunk 7 above)
Comment (IanPhilips) [reply web/pages/group/[...slugs]/index.tsx:430]: yup

File: web/pages/groups.tsx (hunk 8 above)
Comment (IanPhilips) [reply web/pages/groups.tsx:103]: Nice!

File: web/pages/groups.tsx (hunk 9 above)
Comment (IanPhilips) [reply web/pages/groups.tsx:133]: Nice, thanks

File: web/pages/groups.tsx (hunk 10 above)
Comment (IanPhilips) [reply web/pages/groups.tsx:199]: Good idea!

File: web/components/groups/create-group-button.tsx [hunk 20]
```diff
@@ -0,0 +1,140 @@
...
+      setIsSubmitting(false)
+      return true
+    } else {
+      setIsSubmitting(false)
+      return false
+    }
+  }
+
+  return (
+    <ConfirmationButton
```
[the remark this answers is not in the document]
Comment (IanPhilips) [reply web/components/groups/create-group-button.tsx:80]: I will do this....  🔜

**Pull request merged**